aspnetcore/security/authentication/configure-oidc-web-authentication.md

```diff
@@ -0,0 +1,319 @@
+
+[OpenID Connect](https://openid.net/developers/how-connect-works/) can be used to implement authentication in ASP.NET Core applications. The recommended way is to use an OpenID Connect confidential client using the code flow. Using the [Proof Key for Code Exchange by OAuth Public Clients (PKCE)](https://datatracker.ietf.org/doc/html/rfc7636) is recommended for this implementation. Both the application client and the user of the application are authenticated in the confidential flow. The application client uses a client secret or a client assertion to authenticate. 
+
+Public OpenID Connect/OAuth clients are no longer recommended for web applications.
+
+The default flow works as shown in the following diagram:
+
+![OIDC code flow confidential client using PKCE](~/security/authentication/configure-oidc-web-authentication/_static/oidc-confidential-pkce-flow-drawio.png)
+
+OpenID Connect comes in many variations and all server implementations have slightly different parameters and requirements. Some servers don’t support the user info endpoint, some still don’t support PKCE and others require special parameters in the token request. Client assertions can be used instead of client secrets. New standards also exist which add extra security on top of the OpenID Connect Core, for example FAPI, CIBA or DPoP for downstream APIs.
+
+> [!NOTE]
+> From .NET 9, [OAuth 2.0 Pushed Authorization Requests (PAR) RFC 9126](https://datatracker.ietf.org/doc/html/rfc9126) is used per default, if the OpenID Connect server supports this. This is a three step flow and not a two step flow as shown above. (User Info request is an optional step.)
+
+## Create an Open ID Connect code flow client using Razor Pages
+
+The following section shows how to implement an OpenID Connect client in an empty ASP.NET Core Razor page project. The same logic can be applied to any ASP.NET Core web project with only the UI integration being different.
+
+### Add OpenID Connect support
+
+Add the [Microsoft.AspNetCore.Authentication.OpenIdConnect](https://www.nuget.org/packages/Microsoft.AspNetCore.Authentication.OpenIdConnect) Nuget packages to the ASP.NET Core project.
+
+### Setup the OpenID Connect client
+
+Add the authentication to the web application using the builder.Services in the **Program.cs** file. The configuration is dependent on the OpenID Connect server. Each OpenID Connect server requires small differences in the setup.
+
+```csharp
+builder.Services.AddAuthentication(options =>
+{
+	options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+	options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+})
+.AddCookie()
+.AddOpenIdConnect(options =>
+{
+	configuration.GetSection("OpenIDConnectSettings").Bind(options);
+	
+	options.Authority = configuration["OpenIDConnectSettings:Authority"];
+	options.ClientId = configuration["OpenIDConnectSettings:ClientId"];
+	options.ClientSecret = configuration["OpenIDConnectSettings:ClientSecret"];
```

Comment:
```suggestion
    var oidcConfig = configuration.GetSection("OpenIDConnectSettings");

	  options.Authority = oidcConfig["Authority"];
	  options.ClientId = oidcConfig["ClientId"];
	  options.ClientSecret = oidcConfig["ClientSecret"];
```

Isn't the call to `Bind(options)` redundant? I like explicitly setting the properties over just calling `Bind` to show which ones matter. And it doesn't require the reader to understand config binding to get a general idea of what's going on.

Also, I think we try to use 4 spaces for indentation everywhere.

Reply:
fixed